src/data/padics/ring_homs.lean

```diff
@@ -0,0 +1,689 @@
+See `padic_int.zmod_repr` for a version that takes values in `ℕ`
+and works for arbitrary `x : ℤ_[p]`.) -/
+def mod_part : ℤ :=
+(r.num * gcd_a r.denom p) % p
+
+include hp_prime
+
+variable {p}
+lemma mod_part_lt_p : mod_part p r < p :=
+begin
+  convert int.mod_lt _ _,
+  { simp },
+  { exact_mod_cast hp_prime.ne_zero }
+end
+
+lemma mod_part_nonneg : 0 ≤ mod_part p r :=
+int.mod_nonneg _ $ by exact_mod_cast hp_prime.ne_zero
+
+lemma is_unit_denom (r : ℚ) (h : ∥(r : ℚ_[p])∥ ≤ 1) : is_unit (r.denom : ℤ_[p]) :=
+begin
+  rw is_unit_iff,
+  apply le_antisymm (r.denom : ℤ_[p]).2,
+  rw [← not_lt, val_eq_coe, coe_coe],
+  intro norm_denom_lt,
+  have hr : ∥(r * r.denom : ℚ_[p])∥ = ∥(r.num : ℚ_[p])∥,
+  { rw_mod_cast @rat.mul_denom_eq_num r, refl, },
+  rw padic_norm_e.mul at hr,
+  have key : ∥(r.num : ℚ_[p])∥ < 1,
+  { calc _ = _ : hr.symm
+    ... < 1 * 1 : _
+    ... = 1 : mul_one 1,
+    apply mul_lt_mul' h norm_denom_lt (norm_nonneg _) zero_lt_one, },
+  have : ↑p ∣ r.num ∧ (p : ℤ) ∣ r.denom,
+  { simp only [← norm_int_lt_one_iff_dvd, ← padic_norm_e_of_padic_int],
+    norm_cast, exact ⟨key, norm_denom_lt⟩ },
+  apply hp_prime.not_dvd_one,
+  rwa [← r.cop.gcd_eq_one, nat.dvd_gcd_iff, ← int.coe_nat_dvd_left, ← int.coe_nat_dvd],
+end
+
+lemma norm_sub_mod_part_aux (r : ℚ) (h : ∥(r : ℚ_[p])∥ ≤ 1) :
+  ↑p ∣ r.num - r.num * r.denom.gcd_a p % p * ↑(r.denom) :=
+begin
+  rw ← zmod.int_coe_zmod_eq_zero_iff_dvd,
+  simp only [int.cast_coe_nat, zmod.cast_mod_nat p, int.cast_mul, int.cast_sub],
+  have := congr_arg (coe : ℤ → zmod p) (gcd_eq_gcd_ab r.denom p),
+  simp only [int.cast_coe_nat, add_zero, int.cast_add, zmod.cast_self, int.cast_mul, zero_mul] at this,
+  push_cast,
+  rw [mul_right_comm, mul_assoc, ←this],
+  suffices rdcp : r.denom.coprime p,
+  { rw rdcp.gcd_eq_one, simp only [mul_one, cast_one, sub_self], },
+  apply coprime.symm,
+  apply (coprime_or_dvd_of_prime ‹_› _).resolve_right,
+  rw [← int.coe_nat_dvd, ← norm_int_lt_one_iff_dvd, not_lt],
+  apply ge_of_eq,
+  rw ← is_unit_iff,
+  exact is_unit_denom r h,
+end
+
+lemma norm_sub_mod_part (h : ∥(r : ℚ_[p])∥ ≤ 1) : ∥(⟨r,h⟩ - mod_part p r : ℤ_[p])∥ < 1 :=
+begin
+  let n := mod_part p r,
+  by_cases aux : (⟨r,h⟩ - n : ℤ_[p]) = 0,
+  { rw [aux, norm_zero], exact zero_lt_one, },
+  rw [norm_lt_one_iff_dvd, ← (is_unit_denom r h).dvd_mul_right],
+  suffices : ↑p ∣ r.num - n * r.denom,
+  { convert (int.cast_ring_hom ℤ_[p]).map_dvd this,
+    simp only [sub_mul, int.cast_coe_nat, ring_hom.eq_int_cast, int.cast_mul,
+      sub_left_inj, int.cast_sub],
+    apply subtype.coe_injective,
+    simp only [coe_mul, subtype.coe_mk, coe_coe],
+    rw_mod_cast @rat.mul_denom_eq_num r, refl },
+  dsimp [n, mod_part],
+  apply norm_sub_mod_part_aux r h,
+end
+
+lemma exists_mem_range_of_norm_rat_le_one (h : ∥(r : ℚ_[p])∥ ≤ 1) :
+  ∃ n : ℤ, 0 ≤ n ∧ n < p ∧ ∥(⟨r,h⟩ - n : ℤ_[p])∥ < 1 :=
+⟨mod_part p r, mod_part_nonneg _, mod_part_lt_p _, norm_sub_mod_part _ h⟩
+
+end
+
+section
```

Comment:
Uniquely named sections help me to find my way around a file faster. This is the second no-name section. Is it a good idea to name them?

Reply:
In fact, these section breaks don't really correspond to anything!